tests/contrib/patch.py

```diff
@@ -51,151 +56,305 @@ def assert_not_double_wrapped(self, obj):
-        """
-        To ensure that we can thoroughly test the installation/patching of an
-        integration we must be able to unpatch it before importing the library.
+class PatchTestCase(object):
+    """
+    unittest or other test runners will pick up the base test case as a testcase
+    since it inherits from unittest.TestCase unless we wrap it with this empty
+    parent class.
+    """
+    @run_in_subprocess
+    class Base(SubprocessTestCase, PatchMixin):
+        """PatchTestCase provides default test methods to be used for testing
+        common patching logic. Each test method provides a default
+        implementation which will use `self.MODULES` if it exists.
 
-        For example::
+        If `self.MODULES` is not provided and the test method is not overridden,
+        a NotImplementedError will be raised encouraging the user to implement
+        the test.
 
-            ddtrace.patch(redis=True)
-            from ddtrace.contrib.redis import unpatch
-            unpatch()
-            import redis
-            self.assert_not_wrapped(redis.StrictRedis.execute_command)
-            self.assert_not_wrapped(redis.StrictRedis.pipeline)
-            self.assert_not_wrapped(redis.Redis.pipeline)
-            self.assert_not_wrapped(redis.client.BasePipeline.execute)
-            self.assert_not_wrapped(redis.client.BasePipeline.immediate_execute_command)
-        """
-        raise NotImplementedError(self.test_unpatch_before_import.__doc__)
+        Attributes:
+            MODULES: A list of tuples of the form
+                    (module_name, integration_name, unpatch_method)
 
-    def test_unpatch_after_import(self):
-        """
-        To ensure that we can thoroughly test the installation/patching of an
-        integration we must be able to unpatch it after importing the library.
+        Example:
 
-        For example::
+        A simple implementation inheriting this TestCase looks like::
 
-            import redis
             from ddtrace.contrib.redis import unpatch
-            ddtrace.patch(redis=True)
-            unpatch()
-            self.assert_not_wrapped(redis.StrictRedis.execute_command)
-            self.assert_not_wrapped(redis.StrictRedis.pipeline)
-            self.assert_not_wrapped(redis.Redis.pipeline)
-            self.assert_not_wrapped(redis.client.BasePipeline.execute)
-            self.assert_not_wrapped(redis.client.BasePipeline.immediate_execute_command)
-        """
-        raise NotImplementedError(self.test_unpatch_after_import.__doc__)
 
-    def test_unpatch_patch(self):
-        """
-        To ensure that we can thoroughly test the installation/patching of an
-        integration we must be able to unpatch it and then subsequently patch it
-        again.
+            class RedisPatchTestCase(PatchTestCase.Base):
+                # 'redis' module, 'redis' integration, redis unpatch method
+                MODULES = [('redis', 'redis', unpatch)]
```

Comment:
A tuple seems a bit fragile here. However, this should not be changing often or at all. Thoughts?